### ISSUE TITLE
Where are the Revit shared parameters installed?

### ISSUE THREAD

**Issue by Moult** [Autodesk/revit-ifc#461]
In this document https://blogs.autodesk.com/revit/wp-content/uploads/sites/110/2022/02/09/revit-ifc-open-bim-manual-en.pdf

It says: "The most relevant among these are the IFC Shared parameter files which are used to add IFC properties to Revit. These are stored in: C:\ProgramData\Autodesk\ApplicationPlugins\IFC 20xx.bundle"

I don't have this folder on my machine. Where is it installed? I've also checked under Program Files instead of ProgramData and Program Files (X86) and also in %appdata%.

How are users meant to discover these files? Am I missing a set of docs?

**Comment by Thomas84** [comment]
Hi @Moult, 
I have old Revit versions on my machine, and until Revit 2021 the path seems to be correct. But I have downloaded them from https://github.com/Autodesk/revit-ifc/tree/master/Source/RevitIFCTools 
I have looked into the parameters in that files and as mentioned [here](https://github.com/Autodesk/revit-ifc/issues/453), they are not complete, because some Revit specific parameters are missing. With 2023 some parameters will get obsolete, because they are then built-in parameters.

**Comment by Moult** [comment]
Yes indeed, but for users on 2021 or 2022, where should they find the shared param files? From their local install? Download from Github? Which link? From a user perspective just reading the manual it seems unclear.

Also what's the difference between the shared params files here https://github.com/Autodesk/revit-ifc/tree/master/Source/RevitIFCTools ? Type? Type_ALL? Type_ALL_grouped? Which one should a user use?

This is the one I've personally downloaded: https://github.com/Autodesk/revit-ifc/tree/master/Install/Program%20Files%20to%20Install - why are the files in two locations? Which is correct?

**Comment by AngelVelezSosa** [comment]
We are working obsoleting all of this by using the new Parameters Service (in tech preview at the moment).  This will eliminate duplicates and have one source of truth for these parameters moving forward.

**Comment by Moult** [comment]
Is that going to be made available to Revit 2020? 2021? 2022? 2023?

**Comment by Thomas84** [comment]
My guess is, that the "parameter service" is a "cloud users only" feature (as far as I know). If you don't have access to BIM360 / ACC you cannot access the parameter service. A standard Revit user can see that there is a tool "Parameter Service", but cannot access it. 

I hope there will be in the future a better solution for parameters, because the concept of the actual shared parameters is not good. Parameters in general in Revit needs an large update, to be more flexible with selection sets, data ranges and the typical settings for values, with the same functionality that can be done in a database...

**Comment by AngelVelezSosa** [comment]
Keep in mind that everything I say from this point onward is based on my limited knowledge and/or speculation.
I believe that the "parameter service" is a cloud system, similar to bsDD, and will likely continue to be.  It solves the inherent problem of the "shared" part of shared parameters, in that they are not easily globally shared.  You are right that there are two aspects to parameters: (1) their functionality and (2) their reusability.  The parameter service (or bsDD, for that matter) attempts to solve #2.  Work on #1 is outside the scope of this discussion/IFC, although I believe that that will regardless be important and necessary for other reasons, and will get some attention.

**Comment by benwwchen** [comment]
> Is that going to be made available to Revit 2020? 2021? 2022? 2023?
> 
> When will it be available? (i.e. on the app store for end-users?) If it's half a year or more, what is the solution in the meantime?

Same question here after one year. I'm curious about if there are any updates on this matter, either on the new "service" or solution for old Revit versions? Are the shared parameters file valid or fixed for older versions of Revit (e.g., 2021)?  Or where can I find the internal mappings from Revit parameters to IFC properties actually used by the Revit-IFC add-in?

**Comment by AngelVelezSosa** [comment]
The Parameter Service is live in Revit for new versions.  It is not available for older versions.  For older versions, there is the "confusion" of the two locations which will have to be sorted out with the people you are collaborating with, but they should be generally the same anyway.

**Comment by benwwchen** [comment]
> The Parameter Service is live in Revit for new versions. It is not available for older versions. For older versions, there is the "confusion" of the two locations which will have to be sorted out with the people you are collaborating with, but they should be generally the same anyway.

Thanks for the reply. If I'm not mistaken, that means we could use either file for exporting IFC property sets in older versions of Revit? However, I found the naming for the parameters are different in the latest one on GitHub from the one included in the add-in. And many parameters are not prefixed by Pset names so there are still many confusions regarding the IFC shared parameters txt file. The main problem is how to name the parameters so that they are automatically mapped to IFC properties correctly.

Therefore, are there any ways we can find the internal mappings from Revit parameters to IFC properties **actually used by the Revit-IFC add-in**? This way I can find out which properties are actually supported by the add-in and which are required to create mappings to either common property sets or user-defined property sets.

**Comment by Moult** [comment]
I took a look at the Parameter Service but from what I can tell, it still requires an account administrator to load in a bunch of shared parameters (from where? From this repo?) (and then do a huge amount of work to setup what is applicable to what). Is that correct?

In addition, it seems as though the Parameter Service is across the entire ACC account, not per ACC project, this means that projects will get confused as to which parameters apply to them.

**Comment by AngelVelezSosa** [comment]
The IFC parameters are not perfect in terms of proper mapping of IFC parameters to Revit categories (although that said, people "misuse" Revit categories all of the time, so being able to add parameters to novel categories is definitely a feature).  It is still much easier than doing one shared parameter at the time.  As for the rest of it, I'm not a parameter service expert, so can't talk to it.  Still, I don't understand "In addition, it seems as though the Parameter Service is across the entire ACC account, not per ACC project, this means that projects will get confused as to which parameters apply to them." - where is the confusion?

Let me see if I can get an official response on that.  By the way, the old way of setting up a template with the shared parameters still works.

**Comment by AngelVelezSosa** [comment]
> > The Parameter Service is live in Revit for new versions. It is not available for older versions. For older versions, there is the "confusion" of the two locations which will have to be sorted out with the people you are collaborating with, but they should be generally the same anyway.
> 
> Thanks for the reply. If I'm not mistaken, that means we could use either file for exporting IFC property sets in older versions of Revit? However, I found the naming for the parameters are quite different in the latest one on GitHub from the one included in the add-in. And many parameters are not prefixed by Pset names so there are still many confusions regarding the IFC shared parameters txt file. The main problem is **how to name the parameters** so that they are automatically mapped to IFC properties by the Add-in correctly.
> 
> Therefore, are there any ways we can find the internal mappings from Revit parameters to IFC properties **actually used by the Revit-IFC add-in**? This way I can find out which properties are actually supported by the add-in and which are required to create mappings to either common property sets or user-defined property sets.


There are very few parameters mapped, and the values here are generally still true: https://sourceforge.net/p/ifcexporter/wiki/Mapping%20of%20Revit%20parameters%20to%20IFC%20values/

As for the Parameter Service, there is a roadmap here that does contain the issues you mentioned above:
https://aps.autodesk.com/autodesk-parameters-service-roadmap

**Comment by Moult** [comment]
> I don't understand "In addition, it seems as though the Parameter Service is across the entire ACC account, not per ACC project, this means that projects will get confused as to which parameters apply to them." - where is the confusion?

When I tried to use it, it seems as though the parameters are defined at the level of the Autodesk Hub. This means that all projects within that Hub access the same library of parameters. At least it seemed this way in my tests? Is there someone I can reach out to for help (sorry I know this is off-topic and shouldn't be posted here, but not sure where to turn to)

**Comment by lejla-adsk** [comment]
Hi @Moult , this is correct - but you can create collections to keep the parameters specific to one project together (similar to a shared parameters file). 
The user will have to select the collection before selecting parameters, which will help avoid confusion and give the user only the parameters "allowed" in this specific project.
Does this help?

**Comment by Moult** [comment]
@lejla-adsk yeah, collections help, but it was kicked back due to the number of projects (we have lots of projects!) and the potential for leaking contractually sensitive data. We also have air gapped projects which rules out the parameter service. Sorry, I didn't mean to take this issue off topic.

**Comment by gotoarchi** [comment]
Currently you can have the shared parameters in the cloud, in a service currently known in 12/2023 as the Autodesk Parameter Service. You can [read more about it in the article How to enable the autodesk parameter service for Revit](https://goto.archi/blog/post/how-to-enable-the-autodesk-parameter-service).

**Comment by ArtemHnatko** [comment]
Thanks for raising this — it's a valid point of confusion that likely affects more users than just yourself.

After investigation, the `C:\ProgramData\Autodesk\ApplicationPlugins\IFC 20xx.bundle\Contents\` path is only created when the open-source IFC plugin is installed separately (via MSI or Autodesk Desktop App). If you're running Revit's built-in IFC exporter without the open-source plugin installed on top of it, this folder will not exist on your machine.

In the meantime, the Shared Parameters files can be downloaded directly from this repository:
- [`IFC Shared Parameters-RevitIFCBuiltIn_ALL.txt`](https://github.com/Autodesk/revit-ifc/blob/master/Source/RevitIFCTools/IFC%20Shared%20Parameters-RevitIFCBuiltIn_ALL.txt)
- [`IFC Shared Parameters-RevitIFCBuiltIn-Type_ALL.txt`](https://github.com/Autodesk/revit-ifc/blob/master/Source/RevitIFCTools/IFC%20Shared%20Parameters-RevitIFCBuiltIn-Type_ALL.txt)

The parameter naming inconsistencies mentioned in this thread are being tracked separately in #453.

Closing this as a documentation issue — no changes to the exporter are required.
